jax/_src/third_party/scipy/linalg.py

```diff
@@ -0,0 +1,96 @@
+import scipy.linalg
+
+from jax import jit, lax
+from jax._src.numpy import lax_numpy as jnp
+from jax._src.numpy.linalg import norm
+from jax._src.numpy.util import _wraps
+from jax._src.scipy.linalg import rsf2csf, schur
+
+_array_precision = {'i': 1, 'l': 1, 'f': 0, 'd': 1, 'F': 0, 'D': 1}
+
+eps = jnp.finfo(float).eps
+feps = jnp.finfo(jnp.single).eps
+
+def _asarray_square(A):
+  """
+  Wraps asarray with the extra requirement that the input be a square matrix.
+
+  The motivation is that the matfuncs module has real functions that have
+  been lifted to square matrix functions.
+
+  Parameters
+  ----------
+  A : array_like
+      A square matrix.
+
+  Returns
+  -------
+  out : ndarray
+      An ndarray copy or view or other representation of A.
+
+  """
+  A = jnp.asarray(A)
+  if A.ndim != 2 or A.shape[0] != A.shape[1]:
+    raise ValueError('expected square array_like input')
+  return A
+
+@jit
+def _algorithm_11_1_1(F, T):
+  # Algorithm 11.1.1 from Golub and Van Loan "Matrix Computations"
+  N = T.shape[0]
+  minden = jnp.abs(T[0, 0])
+
+  def _outer_loop(p, F_minden):
+    _, F, minden = lax.fori_loop(1, N-p+1, _inner_loop, (p, *F_minden))
+    return F, minden
+
+  def _inner_loop(i, p_F_minden):
+    p, F, minden = p_F_minden
+    j = i+p
+    s = T[i-1, j-1] * (F[j-1, j-1] - F[i-1, i-1])
+    T_row, T_col = T[i-1], T[:, j-1]
+    F_row, F_col = F[i-1], F[:, j-1]
+    ind = (jnp.arange(N) >= i) & (jnp.arange(N) < j-1)
+    val = (jnp.where(ind, T_row, 0) @ jnp.where(ind, F_col, 0) -
+            jnp.where(ind, F_row, 0) @ jnp.where(ind, T_col, 0))
+    s = s + val
+    den = T[j-1, j-1] - T[i-1, i-1]
+    s = jnp.where(den != 0, s / den, s)
+    F = F.at[i-1, j-1].set(s)
+    minden = jnp.minimum(minden, jnp.abs(den))
+    return p, F, minden
+
+  return lax.fori_loop(1, N, _outer_loop, (F, minden))
+
+_FUNM_LAX_DESCRIPTION = """\
+The array returned by :py:func:`jax.scipy.linalg.funm` may differ in dtype
+from the array returned by py:func:`scipy.linalg.funm`. Specifically, in cases
+where all imaginary parts of the array values are close to zero, the SciPy
+function may return a real-valued array, whereas the JAX implementation will
+return a complex-valued array.
+
+Additionally, unlike the SciPy implementation, when ``disp=True`` no warning
+will be printed if the error in the array output is estimated to be large.
```

Comment:
We could possibly use the experimental [checkify](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.checkify.check.html) work to implement something like this, but I'd probably leave that as a TODO.